lib/web_ui/lib/src/engine/platform_dispatcher.dart

```diff
@@ -782,6 +782,13 @@ class EnginePlatformDispatcher extends ui.PlatformDispatcher {
     scheduleFrameCallback!();
   }
 
+  @override
+  void scheduleWarmUpFrame({required ui.VoidCallback beginFrame, required ui.VoidCallback drawFrame}) {
+    Timer.run(beginFrame);
+    // We use timers here to ensure that microtasks flush in between.
+    Timer.run(drawFrame);
```

Comment:
I'm not sure timers is the right choice. Timers have variable precision on the web to protect users from timing attacks. There's no guarantee that a timer will fire before the next vsync. I think we should call `beginFrame` and `drawFrame` back to back synchronously in a microtask. We can flush microtasks using a `Zone`.

However, in the web case, I think there's not much difference between a "warm-up" frame and a regular `requestAnimationFrame`. The actual first frame is rendered by the HTML engine. Flutter is only ever renders subsequent frames. So perhaps on the web this should still be a `requestAnimationFrame`.

Reply:
Actually, we don't flush microtasks between the two on the web anyway: https://github.com/flutter/engine/blob/23d34eaa5007d185eb3edcd97c85d387ded2a665/lib/web_ui/lib/src/engine/initialization.dart#L181

So I'd just call them back-to-back in a microtask, something like this:

```dart
scheduleMicrotask(() {
  beginFrame();
  drawFrame();
});
```

Reply:
This comment is moved from the framework. https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/scheduler/binding.dart#L1028

I agree that Web might want to implement this method differently, especially since Web doesn't even flush the microevents during a regular frame. I'll change it to the following block, is it ok?
```
          beginFrame();
          // TODO(yjbanov): technically Flutter flushes microtasks between
          //                onBeginFrame and onDrawFrame. We don't, which hasn't
          //                been an issue yet, but eventually we'll have to
          //                implement it properly. (Same TODO as in
          //                `initializeEngineServices`).
          endFrame();
```

However, I think keeping the "warm up" behavior (and hence name) here is important. The warm up frame (as documented in [`SchedulerBinding.scheduleWarmUpFrame`](https://github.com/flutter/flutter/blob/f9b3b84d4eb5afdbe9159fb5ac7a7d24a939a3e6/packages/flutter/lib/src/scheduler/binding.dart#L996)) is designed to make the framework build the first layer tree (cold start) sooner. Therefore this is better done as soon as possible instead of in the next vsync/`requestAnimationFrame`. It's ok if the frame is not rendered at all on the Web.

Reply:
(Sorry I made the comment without seeing your reply) Sure! I'll do it. Should I keep the TODO though?

Reply:
Also, should it be `scheduleMicroTask` or `Timer.run` here? According to the doc of `Timer.run`, 
```
  /// Runs the given [callback] asynchronously as soon as possible.
```
I think `Timer.run` would be the better choice.

Reply:
Today, the framework handles warmup-scheduling and it uses timers here even on the web. This change just moves that scheduling logic to dart:ui. Since the bootstrapping logic is somewhat brittle, I would argue we should keep the existing behavior here for now and use Timers for warn-up frames. If we want a change in behavior here for the web and do something else I would suggest doing that in a separate follow-up PR to decouple that from the other changes in case something goes wrong.

Just my 50c, ultimately I leave it up to the web team to make a call.

Reply:
I agree with @goderbauer , lets not combine refactors and logic changes into the same PR.

Reply:
Ah, I didn't know we already did this. In that case I'm assuming that the framework already resolves the race condition that I was worried about (e.g. it will not schedule the second frame until the first frame is done). Yes, I agree let's keep the timers for now.